configuration/directories.go

```diff
@@ -24,12 +24,8 @@ import (
 func HardwareDirectories(settings *viper.Viper) paths.PathList {
 	res := paths.PathList{}
 
-	if IsBundledInDesktopIDE(settings) {
-		ideDir := paths.New(settings.GetString("IDE.Directory"))
-		bundledHardwareDir := ideDir.Join("hardware")
-		if bundledHardwareDir.IsDir() {
-			res.Add(bundledHardwareDir)
-		}
+	for _, bundledHardwareDir := range BuiltinToolsDirectories(Settings) {
+		res.Add(bundledHardwareDir)
```

Comment:
We see here that this is more than a "tools" directory. It is a packages directory, which might contain platforms as well as tools.

I think this makes it a much more valuable feature, but it also makes the use of "tools" chosen here for the Go package API, configuration keys, and descriptions in the documentation inaccurate.

Reply:
Right, let's rename `*Tools*` -> `*Packages*`

Reply:
mmmm, actually no,  you can only have tools inside that path: https://github.com/arduino/arduino-cli/pull/1817/files#diff-bc91ebff69e54330f19e091571eca06927a59b29e43c53f6a4d5c04ee3dba243R41

```
	bundleToolDirs := configuration.BuiltinToolsDirectories(configuration.Settings)
	merr = append(merr, pm.LoadToolsFromBundleDirectories(bundleToolDirs)...)
```
so I think it's correct tools, and it's wrong the local variable name `bundledHardwareDir` that should be `bundledToolsDir`

Reply:
```text
$ arduino-cli version
arduino-cli.exe  Version: git-snapshot Commit: eb6e52af Date: 2022-08-01T11:56:57Z

$ export ARDUINO_DIRECTORIES_DATA="/tmp/empty-directories-data"

$ arduino-cli core list

$ arduino-cli sketch new /tmp/FooSketch
Sketch created in: C:\Users\per\AppData\Local\Temp\FooSketch

$ arduino-cli compile -b arduino:avr:uno /tmp/FooSketch


Used platform Version Path


Error during build: Platform 'arduino:avr' not found: platform not installed
Try running `arduino-cli.exe core install arduino:avr`
```

As expected, no platforms were available because `directories.data` is empty.

Now I set `directories.builtin.tools` to point to a folder that contains packages and try again:

```text
$ export ARDUINO_DIRECTORIES_BUILTIN_TOOLS="C:/Users/per/AppData/Local/Arduino15/packages"

$ arduino-cli core list
ID          Installed Latest Name
arduino:avr 1.8.5     1.8.5  Arduino AVR Boards

$ arduino-cli compile -b arduino:avr:uno /tmp/FooSketch
Sketch uses 444 bytes (1%) of program storage space. Maximum is 32256 bytes.
Global variables use 9 bytes (0%) of dynamic memory, leaving 2039 bytes for local variables. Maximum is 2048 bytes.


Used platform Version Path
arduino:avr   1.8.5   C:\Users\per\AppData\Local\Arduino15\packages\arduino\hardware\avr\1.8.5
```

Voila! It used the platform from `directories.builtin.tools`

Reply:
Well... :-)

turns out that `configuration.HardwareDirectories(...)` should not return the bundled tools directory at all. So I'm removing these two lines. Nice catch @per1234!

Reply:
Another thing to consider is that the "bundled tools" supports a `builtin_tools_versions.txt` that describes the tools packager origin:

```go
	// We scan toolsPath content to find a "builtin_tools_versions.txt", if such file exists
	// then the all the tools are available in the same directory, mixed together, and their
	// name and version are written in the "builtin_tools_versions.txt" file.
	// If no "builtin_tools_versions.txt" is found, then the directory structure is the classic
	// TOOLSPATH/TOOL-NAME/TOOL-VERSION and it will be parsed as such and associated to an
	// "unnamed" packager.

	// TODO: get rid of "builtin_tools_versions.txt"
```

this is a very Arduino IDE 1.8.x legacy, where we have this `builtin_tools_versions.txt` inside an **hardware** folder:

```
$ tree ~/Downloads/arduino-1.8.13/hardware
hardware
├── arduino
│   └── avr
│       ├── boards.txt
│       ├── cores
│       ├── libraries
│       │   ├── EEPROM
│       │   ├── HID
│       │   ├── SoftwareSerial
│       │   ├── SPI
│       │   └── Wire
│       ├── platform.txt
│       ├── programmers.txt
│       └── variants
├── package_index_bundled.json
└── tools
    └── avr
        ├── avr
        │   ├── bin
        │   ├── include
        │   └── lib
        ├── bin
        │   ├── arduinoOTA
        │   ├── avr-c++
        │   ├── avr-c++filt
        │   ├── avr-cpp
        │   ├── avrdude
        │   ├── avr-g++
        │   .....
        ├── builtin_tools_versions.txt
        ,,,
```

I'm not sure that we can remove this "feature" now, but removing the "bundled tools dir" from the "hardware dirs list" should not produce any visible effect on the `arduino-builder` used in the Arduino IDE 1.8.x.

Reply:
https://github.com/arduino/arduino-cli/pull/1817/commits/696ca8212397c0981f11a55c7f02eccc73609bbd

Reply:
>```go
> the directory structure is the classic
>	// TOOLSPATH/TOOL-NAME/TOOL-VERSION and it will be parsed as such and associated to an
>	// "unnamed" packager.
>```

I guess this is the explanation why the wrong `teensy:teensy-compile@1.56.1` tool is used instead of the expected `arduino:avr-gcc@7.3.0-atmel3.6.1-arduino7` while compiling for `arduino:avr:uno` when I set `directories.builtin.tools` to the `packages` subfolder of my normal `directories.data` folder, and shouldn't be cause for concern?:

```text
$ export ARDUINO_DIRECTORIES_DATA="/tmp/empty-directories-data2"

$ export ARDUINO_DIRECTORIES_BUILTIN_TOOLS="C:/Users/per/AppData/Local/Arduino15/packages"

$ ./arduino-cli core install arduino:avr
Tool arduino:avr-gcc@7.3.0-atmel3.6.1-arduino7 already installed
Tool arduino:avrdude@6.3.0-arduino17 already installed
Tool arduino:arduinoOTA@1.3.0 already installed
Downloading packages...
arduino:avr@1.8.5 already downloaded
Installing platform arduino:avr@1.8.5...
Configuring platform....
Platform arduino:avr@1.8.5 installed

$ ./arduino-cli compile -b arduino:avr:uno -v PERTILLISCH-FooSketch/

Using board 'uno' from platform in folder: C:\Users\per\AppData\Local\Temp\empty-directories-data2\packages\arduino\hardware\avr\1.8.5
Using core 'arduino' from platform in folder: C:\Users\per\AppData\Local\Temp\empty-directories-data2\packages\arduino\hardware\avr\1.8.5
Detecting libraries used...
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\teensy\\tools\\teensy-compile\\1.56.1\\avr/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=atmega328p -DF_CPU=16000000L -DARDUINO=10607 -DARDUINO_AVR_UNO -DARDUINO_ARCH_AVR "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\cores\\arduino" "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\variants\\standard" "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\\sketch\\PERTILLISCH-FooSketch.ino.cpp" -o nul
Error while detecting libraries included by C:\Users\per\AppData\Local\Temp\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\sketch\PERTILLISCH-FooSketch.ino.cpp
Generating function prototypes...
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\teensy\\tools\\teensy-compile\\1.56.1\\avr/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=atmega328p -DF_CPU=16000000L -DARDUINO=10607 -DARDUINO_AVR_UNO -DARDUINO_ARCH_AVR "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\cores\\arduino" "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\variants\\standard" "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\\sketch\\PERTILLISCH-FooSketch.ino.cpp" -o "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\\preproc\\ctags_target_for_gcc_minus_e.cpp"
avr-g++: error: device-specs/specs-atmega328p: No such file or directory
```